Ahmad Atef - Feed API Challenge

ahmad-atef opened the pull request.

danillahtin commented:
Hi @ahmad-atef,

This is Danil, I am an instructor at Essential Developer.

The CI build failed.

Please, fix all build failures and let me know once you're done so I can start the review. Thank you 🙂

ahmad-atef commented:
> Hi @ahmad-atef,
> 
> This is Danil, I am an instructor at Essential Developer.
> 
> The CI build failed.
> 
> Please, fix all build failures and let me know once you're done so I can start the review. Thank you 🙂

hey @danillahtin 
Thanks for the review, I guess it should be ready now 🙂

danillahtin commented on FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
```diff
@@ -7,18 +7,28 @@ import Foundation
 	}
 
-	public init(url: URL, client: HTTPClient) {
+	public init(url: URL, client: HTTPClient, mapper: FeedImageMapper = RemoteFeedImageMapper.default) {
 		self.url = url
 		self.client = client
+		self.mapper = mapper
 	}
```
Since there will be no need to create an instance of `RemoteFeedImageMapper` to call a `static` method, we can remove it.
```suggestion
	public init(url: URL, client: HTTPClient) {
		self.url = url
		self.client = client
	}
```

danillahtin commented on FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
```diff
@@ -7,18 +7,28 @@ import Foundation
+			guard let self = self else { return }
+			switch result {
+			case .success((let data, let response)):
+				completion(self.mapper.map(data, response: response))
```
Following from my previous comment, the `map` function could be `static` so there would be no need to create an instance of `RemoteFeedImageMapper ` to call it.

```suggestion
				completion(RemoteFeedImageMapper.map(data, response: response))
```

danillahtin commented on FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
```diff
@@ -7,18 +7,28 @@ import Foundation
+			case .success((let data, let response)):
+				completion(self.mapper.map(data, response: response))
+			case .failure:
+				completion(.failure(RemoteFeedLoader.Error.connectivity))
```
`RemoteFeedLoader` can be omitted to make code shorter

```suggestion
				completion(.failure(Error.connectivity))
```

danillahtin commented on FeedAPIChallenge/Feed API/HTTPClient.swift:
```diff
@@ -9,3 +9,27 @@ public protocol HTTPClient {
 
 	func get(from url: URL, completion: @escaping (Result) -> Void)
 }
+
```
It is not allowed by the instructions to change this file, please revert the change.

danillahtin commented on FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
```diff
@@ -7,18 +7,28 @@ import Foundation
 	public func load(completion: @escaping (FeedLoader.Result) -> Void) {
-		fatalError("Must be implemented")
+		client.get(from: url) { [weak self] result in
+			guard let self = self else { return }
```
Following from my next comment, there would be no reason to unwrap `self`, so it can be a simple nullability check
```suggestion
			guard self != nil else { return }
```

danillahtin commented on FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
```diff
@@ -7,18 +7,28 @@ import Foundation
+		client.get(from: url) { [weak self] result in
+			guard let self = self else { return }
+			switch result {
+			case .success((let data, let response)):
```
The `let` can be defined outside the tuple so the expression can be shorter:

```suggestion
			case let .success((data, response)):
```

danillahtin commented on FeedAPIChallenge/Feed API/FeedMapper.swift:
```diff
@@ -0,0 +1,53 @@
+
+import Foundation
+
+public protocol FeedImageMapper {
+	func map(_ data: Data, response: HTTPURLResponse) -> FeedLoader.Result
+}
```
Since `RemoteFeedLoader` doesn't need an instance conforming to `FeedImageMapper` to access `RemoteFeedImageMapper`'s static method, there's no need for this protocol. I suggest removing this protocol.

```suggestion
```

danillahtin commented on FeedAPIChallenge/Feed API/FeedMapper.swift:
```diff
@@ -0,0 +1,53 @@
+		let image_url: URL
+	}
+
+	var items: [APIFeedImage]?
```
`items` should not be optional according to the api spec, please make it non-optional.
Also there's no need for it to be mutable, please make it a `let` constant

```suggestion
	let items: [APIFeedImage]
```

danillahtin commented on FeedAPIChallenge/Feed API/FeedMapper.swift:
```diff
@@ -0,0 +1,53 @@
+	func map(_ data: Data, response: HTTPURLResponse) -> FeedLoader.Result
+}
+
+public final class RemoteFeedImageMapper: FeedImageMapper {
+	public static let `default` = RemoteFeedImageMapper()
+
+	private static let OK_STATUS_CODE = 200
+
+	public func map(_ data: Data, response: HTTPURLResponse) -> FeedLoader.Result {
+		guard response.statusCode == Self.OK_STATUS_CODE,
+		      let root = try? JSONDecoder().decode(Root.self, from: data)
+		else {
+			return .failure(RemoteFeedLoader.Error.invalidData)
+		}
+		return .success(root.images)
+	}
+}
```
`RemoteFeedImageMapper` is an internal implementation detail of `RemoteFeedLoader`, thus it should not be accessible from outside of the module, please make it `internal`

Following from my other comments, the `map` function could be made `static` so there would be no need to create an instance of `RemoteFeedImageMapper` to call it.

```suggestion
final class RemoteFeedImageMapper: FeedImageMapper {
	private static let OK_STATUS_CODE = 200

	static func map(_ data: Data, response: HTTPURLResponse) -> FeedLoader.Result {
		guard response.statusCode == OK_STATUS_CODE,
		      let root = try? JSONDecoder().decode(Root.self, from: data)
		else {
			return .failure(RemoteFeedLoader.Error.invalidData)
		}
		return .success(root.images)
	}
}
```

danillahtin commented on FeedAPIChallenge/Feed API/FeedMapper.swift:
```diff
@@ -0,0 +1,53 @@
+	}
+}
+
+struct Root: Decodable {
```
`Root` is an internal implementation detail of `RemoteFeedImageMapper` thus it should be private, please make it private

```suggestion
private struct Root: Decodable {
```

danillahtin reviewed (commented):
Hi @ahmad-atef, you almost passed the challenge!

Please, check my review comments and let me know once you're done so I can review it again. 

Thank you 🙂

danillahtin commented on FeedAPIChallenge/Feed API/HTTPClient.swift:
```diff
@@ -9,3 +9,27 @@ public protocol HTTPClient {
 
 	func get(from url: URL, completion: @escaping (Result) -> Void)
 }
+
+public class RemoteHTTPClient: HTTPClient {
+	private let session: URLSession
+
+	init(_ session: URLSession = .shared) {
+		self.session = session
+	}
+
+	enum RemoteHTTPClientError: Error {
+		case noDataNoResponseNoError
+	}
+
+	public func get(from url: URL, completion: @escaping (HTTPClient.Result) -> Void) {
+		session.dataTask(with: url) { data, response, error in
+			if let error = error {
+				completion(.failure(error))
+			} else if let data = data, let response = response as? HTTPURLResponse {
+				completion(.success((data, response)))
+			} else {
+				completion(.failure(RemoteHTTPClientError.noDataNoResponseNoError))
+			}
+		}
+	}
+}
```
It is not allowed by the instructions to change this file, please revert the change.
```suggestion
```

danillahtin reviewed (commented):
Hi @ahmad-atef, almost there!

Please, check my review comments and let me know once you're done so I can review it again. 

Thank you 🙂

ahmad-atef commented:
> Hi @ahmad-atef, almost there!
> 
> Please, check my review comments and let me know once you're done so I can review it again.
> 
> Thank you 🙂

Hi @danillahtin 
I've addressed your comments :)

danillahtin commented on FeedAPIChallenge/Feed API/FeedImageMapper.swift:
```diff
@@ -0,0 +1,46 @@
+
+import Foundation
+
+final class FeedImageMapper {
```
`FeedImageMapper` is defined as a class, so we can instantiate it as `FeedImageMapper()`. However, all its methods are static - it has no instance methods or properties.

It may be confusing for other developers to be able to instantiate a class but then not be able to do anything with it.

As you'll see in the following lectures, you can make the initializer private in such cases since the `FeedImageMapper` is just a namespace for static methods, with no instance-specific behavior.

Or you can make it an enum with no cases, which can never be instantiated!

Both solutions will disallow instantiation at compile time so that `FeedImageMapper()` won't compile ✅

danillahtin reviewed (commented):
Hi @ahmad-atef,

Congratulations, you've passed the challenge 🥇

All tests are passing, so your implementation complies with the specs.

I left one review comment. Feel free to push changes.

Great job! ✅
